// @zoralabs/nft-hooks@1.2.1/dist/backends/zdk/transformUtils/transformV3AskEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformV3AskEvent = void 0;
const queries_sdk_1 = require("@zoralabs/zdk/dist/queries/queries-sdk");
const types_1 = require("../../../types");
function transformV3AskEvent({ event, at, blockInfo, price, eventType, }) {
    let fixedPriceEvent = undefined;
    switch (event.v3AskEventType) {
        case queries_sdk_1.V3AskEventType.V3AskCreated:
            fixedPriceEvent = types_1.FIXED_PRICE_EVENT_TYPES.FIXED_PRICE_CREATED;
            break;
        case queries_sdk_1.V3AskEventType.V3AskCanceled:
            fixedPriceEvent = types_1.FIXED_PRICE_EVENT_TYPES.FIXED_PRICE_CANCELLED;
            break;
        case queries_sdk_1.V3AskEventType.V3AskPriceUpdated:
            fixedPriceEvent = types_1.FIXED_PRICE_EVENT_TYPES.FIXED_PRICE_UPDATED;
            break;
        case queries_sdk_1.V3AskEventType.V3AskFilled:
            fixedPriceEvent = types_1.FIXED_PRICE_EVENT_TYPES.FIXED_PRICE_FILLED;
            break;
    }
    if (fixedPriceEvent) {
        return {
            at,
            ...price,
            blockInfo,
            eventType,
            sender: event.address,
            marketAddress: event.collectionAddress,
            event: fixedPriceEvent,
            side: types_1.FIXED_SIDE_TYPES.ASK,
            raw: {
                source: types_1.FIXED_PRICE_MARKET_SOURCES.ZORA_ASK_V1,
                data: event,
            },
        };
    }
    return undefined;
}
exports.transformV3AskEvent = transformV3AskEvent;
