// @zoralabs/nft-hooks@1.2.1/dist/backends/zdk/transformUtils/transformV2AuctionEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformV2AuctionEvent = void 0;
const types_1 = require("../../../types");
function transformV2AuctionEvent({ event, at, blockInfo, raw, }) {
    if (event.properties.__typename === 'V2AuctionBidEventProperties') {
        return {
            at,
            blockInfo,
            event: types_1.AUCTION_EVENT_TYPES.AUCTION_BID,
            eventType: types_1.TOKEN_TRANSFER_EVENT_CONTEXT_TYPES.TOKEN_MARKET_EVENT,
            sender: event.properties.sender,
            marketAddress: event.collectionAddress,
            amount: parseInt(event.properties.value),
            raw: {
                source: types_1.AUCTION_SOURCE_TYPES.ZORA_RESERVE_V2,
                raw,
            },
        };
    }
    if (event.properties.__typename === 'V2AuctionCanceledEventProperties') {
        return {
            at,
            blockInfo,
            event: types_1.AUCTION_EVENT_TYPES.AUCTION_CANCELLED,
            eventType: types_1.TOKEN_TRANSFER_EVENT_CONTEXT_TYPES.TOKEN_MARKET_EVENT,
            sender: event.properties.tokenOwner,
            marketAddress: event.collectionAddress,
            raw: {
                source: types_1.AUCTION_SOURCE_TYPES.ZORA_RESERVE_V2,
                raw,
            },
        };
    }
    if (event.properties.__typename === 'V2AuctionCreatedEventProperties') {
        return {
            at,
            blockInfo,
            event: types_1.AUCTION_EVENT_TYPES.AUCTION_CREATED,
            sender: event.properties.tokenOwner,
            marketAddress: event.collectionAddress,
            eventType: types_1.TOKEN_TRANSFER_EVENT_CONTEXT_TYPES.TOKEN_MARKET_EVENT,
            raw: {
                source: types_1.AUCTION_SOURCE_TYPES.ZORA_RESERVE_V2,
                raw,
            },
        };
    }
    if (event.properties.__typename === 'V2AuctionEndedEventProperties') {
        return {
            at,
            blockInfo,
            winner: event.properties.winner,
            event: types_1.AUCTION_EVENT_TYPES.AUCTION_ENDED,
            sender: event.properties.tokenOwner,
            marketAddress: event.collectionAddress,
            eventType: types_1.TOKEN_TRANSFER_EVENT_CONTEXT_TYPES.TOKEN_MARKET_EVENT,
            raw: {
                source: types_1.AUCTION_SOURCE_TYPES.ZORA_RESERVE_V2,
                raw,
            },
        };
    }
    return undefined;
}
exports.transformV2AuctionEvent = transformV2AuctionEvent;
