// @zoralabs/nft-hooks/dist/backends/zdk/transformUtils/transformMarketEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformMarketEvent = void 0;
const extractPrice_1 = require("../utils/extractPrice");
const types_1 = require("../../../types");
const dateToISO_1 = require("../utils/dateToISO");
const transformV2AuctionEvent_1 = require("./transformV2AuctionEvent");
const transformV1MarketEvent_1 = require("./transformV1MarketEvent");
const transformTransferEvent_1 = require("./transformTransferEvent");
const transformMintEvent_1 = require("./transformMintEvent");
const transformV3AskEvent_1 = require("./transformV3AskEvent");
function transformMarketEvent(e) {
    const at = {
        timestamp: (0, dateToISO_1.dateToISO)(e.transactionInfo.blockTimestamp),
        blockNumber: e.transactionInfo.blockNumber,
        transactionHash: e.transactionInfo.transactionHash || undefined,
    };
    const blockInfo = {
        timestamp: e.transactionInfo.blockTimestamp,
        blockNumber: e.transactionInfo.blockNumber,
    };
    const price = (0, extractPrice_1.extractPrice)(e);
    let eventType = types_1.TOKEN_TRANSFER_EVENT_CONTEXT_TYPES.TOKEN_MARKET_EVENT;
    switch (e.properties.__typename) {
        case 'MintEvent':
            eventType = types_1.TOKEN_TRANSFER_EVENT_CONTEXT_TYPES.TOKEN_TRANSFER_EVENT;
            return (0, transformMintEvent_1.transformMintEvent)({
                event: e.properties,
                at,
                blockInfo,
                price,
                eventType,
            });
        case 'TransferEvent':
            eventType = types_1.TOKEN_TRANSFER_EVENT_CONTEXT_TYPES.TOKEN_TRANSFER_EVENT;
            return (0, transformTransferEvent_1.transformTransferEvent)({
                event: e.properties,
                at,
                blockInfo,
                price,
                raw: e,
                eventType,
            });
        case 'V1MarketEvent':
            eventType = types_1.TOKEN_TRANSFER_EVENT_CONTEXT_TYPES.TOKEN_MARKET_EVENT;
            return (0, transformV1MarketEvent_1.transformV1MarketEvent)({
                event: e.properties,
                at,
                blockInfo,
                price,
                eventType,
            });
        case 'V2AuctionEvent':
            eventType = types_1.TOKEN_TRANSFER_EVENT_CONTEXT_TYPES.TOKEN_MARKET_EVENT;
            return (0, transformV2AuctionEvent_1.transformV2AuctionEvent)({
                event: e.properties,
                at,
                blockInfo,
                price,
                raw: e,
                eventType,
            });
        case 'V3AskEvent':
            eventType = types_1.TOKEN_TRANSFER_EVENT_CONTEXT_TYPES.TOKEN_MARKET_EVENT;
            return (0, transformV3AskEvent_1.transformV3AskEvent)({
                event: e.properties,
                at,
                blockInfo,
                price,
                raw: e,
                eventType,
            });
        case 'ApprovalEvent':
        case 'Sale':
        default:
            return undefined;
    }
}
exports.transformMarketEvent = transformMarketEvent;
