// @zoralabs/nft-hooks/dist/backends/zdk/transformUtils/transformMarkets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformMarkets = void 0;
const queries_sdk_1 = require("@zoralabs/zdk/dist/queries/queries-sdk");
const types_1 = require("../../../types");
const dateToISO_1 = require("../utils/dateToISO");
const getStandardMarketData_1 = require("../utils/getStandardMarketData");
function transformMarkets(markets) {
    const getReserveAuctionStatus = (status) => {
        if (status === 'ACTIVE') {
            return types_1.MARKET_INFO_STATUSES.ACTIVE;
        }
        if (status === 'CANCELED') {
            return types_1.MARKET_INFO_STATUSES.CANCELED;
        }
        if (status === 'COMPLETED') {
            return types_1.MARKET_INFO_STATUSES.COMPLETE;
        }
        return types_1.MARKET_INFO_STATUSES.UNKNOWN;
    };
    const getV1MarketFixedPriceStatus = (status) => {
        if (status === 'ACTIVE') {
            return types_1.MARKET_INFO_STATUSES.ACTIVE;
        }
        if (status === 'CANCELED') {
            return types_1.MARKET_INFO_STATUSES.CANCELED;
        }
        if (status === 'COMPLETED') {
            return types_1.MARKET_INFO_STATUSES.COMPLETE;
        }
        return types_1.MARKET_INFO_STATUSES.UNKNOWN;
    };
    const getV3AskStatus = (status) => {
        if (status === 'ACTIVE') {
            return types_1.MARKET_INFO_STATUSES.ACTIVE;
        }
        if (status === 'CANCELED') {
            return types_1.MARKET_INFO_STATUSES.CANCELED;
        }
        if (status === 'COMPLETED') {
            return types_1.MARKET_INFO_STATUSES.COMPLETE;
        }
        return types_1.MARKET_INFO_STATUSES.UNKNOWN;
    };
    const marketResponse = [];
    markets.forEach((market) => {
        var _a, _b, _c, _d, _e, _f, _g;
        if (market.marketType === queries_sdk_1.MarketType.V1Ask &&
            ((_a = market === null || market === void 0 ? void 0 : market.properties) === null || _a === void 0 ? void 0 : _a.__typename) === 'V1Ask') {
            marketResponse.push({
                type: types_1.MARKET_TYPES.FIXED_PRICE,
                source: types_1.FIXED_PRICE_MARKET_SOURCES.ZORA_ASK_V1,
                side: types_1.FIXED_SIDE_TYPES.ASK,
                // TODO(iain): fix naming
                status: getV1MarketFixedPriceStatus((_b = market === null || market === void 0 ? void 0 : market.properties) === null || _b === void 0 ? void 0 : _b.v1AskStatus),
                ...(0, getStandardMarketData_1.getStandardMarketData)({ market, amount: market.properties.amount }),
            });
        }
        if (market.marketType === queries_sdk_1.MarketType.V1Offer &&
            ((_c = market === null || market === void 0 ? void 0 : market.properties) === null || _c === void 0 ? void 0 : _c.__typename) === 'V1Offer') {
            marketResponse.push({
                type: types_1.MARKET_TYPES.FIXED_PRICE,
                source: types_1.FIXED_PRICE_MARKET_SOURCES.ZORA_ASK_V1,
                side: types_1.FIXED_SIDE_TYPES.OFFER,
                status: getV1MarketFixedPriceStatus(market.properties.v1OfferStatus),
                ...(0, getStandardMarketData_1.getStandardMarketData)({ market, amount: market.properties.amount }),
            });
        }
        if (market.marketType === queries_sdk_1.MarketType.V2Auction &&
            ((_d = market === null || market === void 0 ? void 0 : market.properties) === null || _d === void 0 ? void 0 : _d.__typename) === 'V2Auction') {
            const expiresAt = market.properties.estimatedExpirationTime;
            marketResponse.push({
                type: types_1.MARKET_TYPES.AUCTION,
                source: types_1.AUCTION_SOURCE_TYPES.ZORA_RESERVE_V2,
                status: getReserveAuctionStatus(market.properties.v2AuctionStatus),
                auctionId: market.properties.auctionId,
                // Duration shouldn't be able to overflow
                duration: parseInt(market.properties.duration, 10),
                startedAt: market.properties.firstBidTime
                    ? {
                        timestamp: market.properties.firstBidTime,
                    }
                    : undefined,
                bids: [],
                endsAt: market.properties.firstBidTime
                    ? {
                        timestamp: (0, dateToISO_1.dateToISO)(expiresAt),
                    }
                    : undefined,
                currentBid: market.properties.highestBidder && market.properties.highestBidPrice
                    ? {
                        creator: market.properties.highestBidder,
                        created: {
                            timestamp: (0, dateToISO_1.dateToISO)(market.transactionInfo.blockTimestamp),
                        },
                        amount: {
                            usd: market.properties.highestBidPrice.usdcPrice
                                ? {
                                    value: (_e = market.properties.highestBidPrice.usdcPrice) === null || _e === void 0 ? void 0 : _e.decimal,
                                    raw: (_f = market.properties.highestBidPrice.usdcPrice) === null || _f === void 0 ? void 0 : _f.raw,
                                    decimals: 18,
                                }
                                : undefined,
                            eth: market.properties.highestBidPrice.nativePrice
                                ? {
                                    value: market.properties.highestBidPrice.nativePrice.decimal,
                                    raw: market.properties.highestBidPrice.nativePrice.raw,
                                    decimals: 18,
                                }
                                : undefined,
                            amount: {
                                raw: market.properties.highestBidPrice.nativePrice.raw,
                                value: market.properties.highestBidPrice.nativePrice.decimal,
                                decimals: market.properties.highestBidPrice.nativePrice.currency.decimals ||
                                    undefined,
                            },
                            symbol: market.properties.highestBidPrice.nativePrice.currency.name,
                            name: market.properties.highestBidPrice.nativePrice.currency.name,
                            address: market.properties.highestBidPrice.nativePrice.currency.address,
                        },
                    }
                    : undefined,
                ...(0, getStandardMarketData_1.getStandardMarketData)({
                    market,
                    amount: market.properties.highestBidPrice || market.properties.reservePrice,
                }),
            });
        }
        if (market.marketType === queries_sdk_1.MarketType.V3Ask &&
            ((_g = market === null || market === void 0 ? void 0 : market.properties) === null || _g === void 0 ? void 0 : _g.__typename) === 'V3Ask') {
            marketResponse.push({
                type: types_1.MARKET_TYPES.FIXED_PRICE,
                source: types_1.FIXED_PRICE_MARKET_SOURCES.ZORA_ASK_V3,
                side: types_1.FIXED_SIDE_TYPES.ASK,
                status: getV3AskStatus(market.properties.v3AskStatus),
                ...(0, getStandardMarketData_1.getStandardMarketData)({ market, amount: market.properties.askPrice }),
            });
        }
    });
    return marketResponse;
}
exports.transformMarkets = transformMarkets;
